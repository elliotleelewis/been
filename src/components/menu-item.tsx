import { useSetAtom } from 'jotai';
import { type FC, memo, useCallback } from 'react';
import type { Country } from '../models/country';
import { addCountryAtom, removeCountryAtom } from '../state/atoms.ts';

interface Props {
	country: Country;
}

export const MenuItem: FC<Props> = memo(({ country }) => {
	const addCountry = useSetAtom(addCountryAtom);
	const removeCountry = useSetAtom(removeCountryAtom);

	const toggleCountry = useCallback(() => {
		if (country.selected) {
			removeCountry(country.iso3166);
		} else {
			addCountry(country.iso3166);
		}
	}, [country, addCountry, removeCountry]);

	return (
		<li className="hover:bg-zinc-50 dark:hover:bg-zinc-800">
			<label className="flex items-center px-4 py-1">
				<input
					id={country.iso3166}
					className="my-0 me-2 size-4 rounded border-zinc-400 bg-zinc-50 text-primary focus:ring-2 focus:ring-primary/50 active:ring-primary dark:border-zinc-600 dark:bg-zinc-900"
					type="checkbox"
					checked={country.selected}
					onChange={toggleCountry}
				/>
				<span>{country.name}</span>
			</label>
		</li>
	);
});
MenuItem.displayName = 'MenuItem';
