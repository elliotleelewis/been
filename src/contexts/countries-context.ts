import { createContext, useContext } from 'react';

import type { Country } from '../models/country';
import type { Region } from '../models/region';

export interface CountriesContextType {
	countries: readonly Country[];
	regions: readonly Region[];
	focus: Country | null;
	addCountry: (countryCode: string) => void;
	removeCountry: (countryCode: string) => void;
	clearCountries: () => void;
}

export const CountriesContext = createContext<CountriesContextType | undefined>(
	undefined,
);

export const useCountries = (): CountriesContextType => {
	const context = useContext(CountriesContext);
	if (!context) {
		throw new Error('useCountries must be used within a CountriesProvider');
	}
	return context;
};
