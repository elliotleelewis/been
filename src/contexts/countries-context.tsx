import {
	type FC,
	type ReactNode,
	createContext,
	useCallback,
	useContext,
	useEffect,
	useMemo,
	useState,
} from 'react';

import { useLocalStorage } from '../hooks/use-local-storage';
import { type Country } from '../models/country';
import { type Region } from '../models/region';
import { regionalizer } from '../utils/regionalizer';

const COUNTRIES_STORAGE_KEY = 'APP_COUNTRIES';

interface CountriesContextType {
	countries: readonly Country[];
	regions: readonly Region[];
	focus: string | null;
	addCountry: (countryCode: string) => void;
	removeCountry: (countryCode: string) => void;
	clearCountries: () => void;
}

export const CountriesContext = createContext<CountriesContextType | undefined>(
	undefined,
);

export const useCountries = (): CountriesContextType => {
	const context = useContext(CountriesContext);
	if (!context) {
		throw new Error('useCountries must be used within a CountriesProvider');
	}
	return context;
};

interface Props {
	data: readonly Country[];
	children: ReactNode;
}

export const CountriesProvider: FC<Props> = ({ data, children }) => {
	const localStorage = useLocalStorage();
	const [selectedCountries, setSelectedCountries] = useState<string[]>(() => {
		const item = localStorage.getItem(COUNTRIES_STORAGE_KEY);
		return item ? (JSON.parse(item) as string[]) : [];
	});
	const [focus, setFocus] = useState<string | null>(null);

	useEffect(() => {
		localStorage.setItem(
			COUNTRIES_STORAGE_KEY,
			JSON.stringify(selectedCountries),
		);
	}, [selectedCountries, localStorage]);

	const countries = useMemo((): readonly Country[] => {
		return data.map((c) => ({
			...c,
			selected: selectedCountries.includes(c.iso3166),
		}));
	}, [data, selectedCountries]);

	const regions = useMemo((): readonly Region[] => {
		return regionalizer(countries);
	}, [countries]);

	const addCountry = useCallback((countryCode: string) => {
		setSelectedCountries((prevCountries) => {
			if (!prevCountries.includes(countryCode)) {
				setFocus(countryCode);
				return [...prevCountries, countryCode];
			}
			return prevCountries;
		});
	}, []);

	const removeCountry = useCallback((countryCode: string) => {
		setSelectedCountries((prevCountries) => {
			const newCountries = prevCountries.filter(
				(code) => code !== countryCode,
			);
			return newCountries;
		});
	}, []);

	const clearCountries = useCallback(() => {
		setSelectedCountries([]);
	}, []);

	const value = useMemo(
		() => ({
			countries,
			regions,
			focus,
			addCountry,
			removeCountry,
			clearCountries,
		}),
		[countries, regions, focus, addCountry, removeCountry, clearCountries],
	);

	return (
		<CountriesContext.Provider value={value}>
			{children}
		</CountriesContext.Provider>
	);
};
