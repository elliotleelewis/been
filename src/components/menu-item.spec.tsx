import { render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { CountriesContext } from '../contexts/countries-context.ts';
import { MenuItem } from './menu-item.ts';

describe('MenuItem', () => {
	it('should render', () => {
		const result = render(
			<CountriesContext.Provider
				value={{
					countries: [],
					regions: [],
					focus: null,
					addCountry: vi.fn(),
					removeCountry: vi.fn(),
					clearCountries: vi.fn(),
				}}
			>
				<MenuItem
					country={{
						name: 'United Kingdom',
						iso3166: 'GB',
						region: 'Europe',
					}}
				/>
			</CountriesContext.Provider>,
		);

		expect(result.asFragment()).toMatchSnapshot();
	});
});
