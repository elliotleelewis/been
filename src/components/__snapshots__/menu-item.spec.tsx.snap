// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MenuItem > should render 1`] = `
<DocumentFragment>
  <li
    class="hover:bg-zinc-50 dark:hover:bg-zinc-800"
  >
    <label
      class="flex items-center px-4 py-1"
    >
      <input
        class="my-0 me-2 size-4 rounded border-zinc-400 bg-zinc-50 text-primary focus:ring-2 focus:ring-primary/50 active:ring-primary dark:border-zinc-600 dark:bg-zinc-900"
        id="GB"
        type="checkbox"
      />
      <span>
        United Kingdom
      </span>
    </label>
  </li>
</DocumentFragment>
`;
